Iterables (#179)
javascript-tutorial/es.javascript.info: 1-js/05-data-types/06-iterable/article.md

while (true) {
  let result = iterator.next();
  if (result.done) break;
  alert(result.value); // retorna los caracteres uno por uno
}
```

Rara vez se necesita esto, pero nos da más control sobre el proceso que `for..of`. Por ejemplo, podemos dividir el proceso de iteración: iterar un poco, luego parar, hacer otra cosa y luego continuar.
 
## Iterables y array-likes [#array-like]

Hay dos términos oficiales que se parecen, pero son muy diferentes. Asegúrese de comprenderlos bien para evitar confusiones.
 
- *Iterables* son objetos que implementan el método `Symbol.iterator`, como se describió anteriormente.
- *Array-likes* son objetos que tienen índices y `longitud` o *length*, por lo que se ven como matrices.

Naturalmente, estas propiedades pueden combinarse. Por ejemplo, las cadenas son iterables (`for..of` funciona en ellas) y tienen forma de matriz (tienen índices numéricos y` longitud` o *length*).

Pero un iterable puede no tener forma de matriz. Y viceversa, un tipo de matriz puede no ser iterable.

Por ejemplo,`range` en el ejemplo anterior es iterable, pero no como una matriz, porque no tiene propiedades indexadas ni`longitud` o *length*.

> Y aquí está el objeto que tiene forma de matriz, pero no es iterable:

Reviewer: ```suggestion

Y aquí está el objeto que tiene forma de matriz, pero no es iterable:
```

```js run
let arrayLike = { // tiene índices y longitud => array-like
  0: "Hola",
  1: "Mundo",
  length: 2
};

*!*

> // Error (arrayLike no es un iterable)

Reviewer: ```suggestion
// Error (sin Symbol.iterator)
```

for (let item of arrayLike) {}
*/!*
```

> ¿Qué tienen en común? Tanto los iterables como los array-like generalmente son *no matrices*, no tienen "push", "pop", etc. Eso genera inconvenientes si tenemos un objeto así y queremos trabajar con él como con una matriz.

Reviewer: ```suggestion
Tanto los iterables como los array-like generalmente no son *matrices*, no tienen "push", "pop", etc. Eso es bastante inconveniente si tenemos un objeto de este tipo y queremos trabajar con él como con una matriz. P.ej. nos gustaría trabajar con `range` utilizando métodos de matriz. ¿Cómo lograr eso?
```

## Array.from

> Existe un método universal [Array.from](https://developer.mozilla.org/es/docs/Web/JavaScript/Referencia/Objetos_globales/Array/from) que toma un valor iterable o similar a una matriz y crea una matriz ¨real¨ a partir de él. De esta manera podemos llamar y usar métodos que pertenecen a una matriz.

Reviewer: ```suggestion
Existe un método universal [Array.from](https://developer.mozilla.org/es/docs/Web/JavaScript/Referencia/Objetos_globales/Array/from) que toma un valor iterable o similar a una matriz y crea un `Array` ¨real¨ a partir de él. De esta manera podemos llamar y usar métodos que pertenecen a una matriz.
```

Por ejemplo:

```js run
let arrayLike = {
  0: "Hola",
  1: "Mundo",
  length: 2
};

*!*
let arr = Array.from(arrayLike); // (*)
*/!*
alert(arr.pop()); // Mundo (el método pop funciona)
```

`Array.from` en la línea `(*)` toma el objeto, lo examina por ser iterable o similar a una matriz, luego crea una nueva matriz y copia allí todos los elementos.

Lo mismo sucede para un iterable:

```js
// suponiendo que range se toma del ejemplo anterior
let arr = Array.from(range);
alert(arr); // 1,2,3,4,5 (la conversión de matriz a cadena funciona)
```

> La sintaxis completa para `Array.from` permite proporcionar una función opcional de "mapeo" :

Reviewer: ```suggestion
La sintaxis completa para `Array.from` también nos permite proporcionar una función opcional de "mapeo":
```

> 

Reviewer: ```suggestion
```

```js
Array.from(obj[, mapFn, thisArg])
```

> El segundo argumento `mapFn` debería ser la función que se aplicará a cada elemento antes de agregarlo a la matriz, y` thisArg` permite establecerlo mediante el `this`.

Reviewer: ```suggestion
El segundo argumento opcional `mapFn` puede ser una función que se aplicará a cada elemento antes de agregarlo a la matriz, y `thisArg` permite establecer el `this` para ello.
```

alert(chars.length); // 2
```

> A diferencia de `str.split`,`Array.from` se basa en la naturaleza iterable de la cadena y, por lo tanto, al igual que`for..of`, funciona correctamente con pares sustitutos.

Reviewer: ```suggestion
A diferencia de `str.split`, `Array.from` se basa en la naturaleza iterable de la cadena y, por lo tanto, al igual que `for..of`, funciona correctamente con pares sustitutos.
```

```

## Resumen

Los objetos que se pueden usar en `for..of` se denominan *iterables*.

- Técnicamente, los iterables deben implementar el método llamado `Symbol.iterator`.
    - El resultado de `obj[Symbol.iterator]` se llama *iterador*. Maneja el proceso de iteración adicional.
    - Un iterador debe tener el método llamado `next()` que devuelve un objeto `{done: Boolean, value: any}`, donde `done: true` denota el final de la iteración, de lo contrario, `value` es el siguiente valor.
- El método `Symbol.iterator` se llama automáticamente por `for..of`, pero también podemos hacerlo directamente.
- Los iterables integrados, como cadenas o matrices, también implementan `Symbol.iterator`.
- El iterador de cadena funciona con pares sustitutos.
 

Los objetos que tienen propiedades indexadas y `longitud` o *length* se llaman *array-like*. Dichos objetos también pueden tener otras propiedades y métodos, pero carecen de los métodos integrados de las matrices.

Si miramos dentro de la especificación, veremos que la mayoría de los métodos incorporados suponen que funcionan con iterables o array-likes en lugar de matrices "reales", porque eso es más abstracto.

> `Array.from (obj[, mapFn, thisArg])` crea un verdadero `Array` de un` obj` iterable o array-like, y luego podemos usar métodos de matriz en él. Los argumentos opcionales `mapFn` y` thisArg` nos permiten aplicar una función a cada elemento.

Reviewer: ```suggestion
`Array.from (obj[, mapFn, thisArg])` crea un verdadero `Array` de un `obj` iterable o array-like, y luego podemos usar métodos de matriz en él. Los argumentos opcionales `mapFn` y `thisArg` nos permiten aplicar una función a cada elemento.
```

parece que los ` nos cambian los espacios
